### DISCUSSION TITLE
How to add additional renderers to the main loop

### DISCUSSION THREAD

**Discussion by freshcutdevelopment** [iTowns/itowns#1964] [Q&A]
I'm currently using iTowns to render point clouds. It's working great.

As a next step I'd like to add measurement using the Raycaster, something like this?

https://sbcode.net/threejs/measurements

However, I can't see an easy way of adding additional renderers to the main loop. I need to be able to do something like this:

```js
function animate() {
    requestAnimationFrame(animate)

    controls.update()

    render()

    stats.update()
}

function render() {
    labelRenderer.render(scene, camera)
    renderer.render(scene, camera)
}

```

Is this achievable?  If somebody could point me in the right direction that'd be very helpful.

**Comment by jailln** [comment]
Hi,
You can add a `FrameRequester` to the view with [addFrameRequester](http://www.itowns-project.org/itowns/docs/#api/View/View). You can specify `when` this `FrameRequester` will be executed with one of the [main loop events](https://github.com/iTowns/itowns/blob/master/src/Core/MainLoop.js#L6). For instance, to render the labels after the scene has been rendered, you can write something like: 

```
function renderLabel() {
    labelRenderer.render(view.scene, view.camera.camera3D);
}
view.addFrameRequester(MAIN_LOOP_EVENTS.AFTER_RENDER, renderLabel);
```

A distance measurement tool would benefit itowns community, would you be keen to implement it as an itowns widget and push it to the main repository? :) (widgets in itowns are applicative features packaged separately from itowns that are configurable and can be used by other people building applications with itowns, an example is [the minimap widget](https://www.itowns-project.org/itowns/examples/#widgets_minimap).
 
I openned [an issue](https://github.com/iTowns/itowns/issues/1868) a few months ago to add two widgets: an elevation measurement tool, which [I already started to implement](https://github.com/iTowns/itowns/pull/1881) and a distance measurement tool like you want to implement.
